### PR TITLE
Basic input

### PR DESCRIPTION
**Pull request by JoelEinbinder**
#43 
Adds
```javascript
.click(selector)
.focus(selector)
.type(text)
```

### CODE REVIEW

File: test/assets/input/textarea.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Button test</title>
```
Comment (aslushnikov) [inline test/assets/input/textarea.html]: text area test?

File: lib/Page.js [hunk 2]
```diff
@@ -522,6 +522,88 @@ class Page extends EventEmitter {
   async close() {
     await this._client.dispose();
   }
+
+  /**
+   * @param {string} selector
+   * @param {!Promise<number>}
+   */
+  async _querySelector(selector) {
+    let documentNode = await this._client.send('DOM.getDocument', {
```
Comment (aslushnikov) [inline lib/Page.js]: we should query for the document only once in the very beginning

File: lib/Page.js [hunk 3]
```diff
@@ -522,6 +522,88 @@ class Page extends EventEmitter {
...
+
+  /**
+   * @param {string} selector
+   * @param {!Promise}
+   */
+  async click(selector) {
+    let boxModel = (await this._client.send('DOM.getBoxModel', {
+      nodeId: await this._querySelector(selector)
+    })).model.content;
+    let x = Math.round((boxModel[0] + boxModel[0]) / 2);
```
Comment (aslushnikov) [inline lib/Page.js]: this feels like an error

File: lib/Page.js [hunk 4]
```diff
@@ -522,6 +522,88 @@ class Page extends EventEmitter {
...
+    let y = Math.round((boxModel[1] + boxModel[1]) / 2);
+
+    await this._client.send('Input.dispatchMouseEvent', {
```
Comment (aslushnikov) [inline lib/Page.js]: you can await only the last one due to the serial nature of the protocol

File: lib/Page.js [hunk 5]
```diff
@@ -522,6 +522,88 @@ class Page extends EventEmitter {
...
+      clickCount: 1
+    });
+  }
+
+  /**
+   * @param {string} selector
+   * @param {!Promise}
+   */
+  async focus(selector) {
+    await this.evaluate(selector => {
```
Comment (aslushnikov) [inline lib/Page.js]: let's use DOM.querySelector and DOM.focus

File: lib/Page.js [hunk 6]
```diff
@@ -522,6 +522,88 @@ class Page extends EventEmitter {
...
+
+  /**
+   * @param {string} text
+   * @param {!Promise}
+   */
+  async type(text) {
+    for (let i = 0; i < text.length; i++) {
+      let code = text.charCodeAt(i);
+      let char = text.charAt(i);
+      await this._client.send('Input.dispatchKeyEvent', {
```
Comment (aslushnikov) [inline lib/Page.js]: you'd probably speed this up considerably if you were to wait only for the *very* last one

File: test/assets/input/textarea.html [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  </head>
+  <body>
+    <textarea></textarea>
+    <script>
+      window.result = "";
```
Comment (aslushnikov) [inline test/assets/input/textarea.html]: ''

**Review by aslushnikov — CHANGES_REQUESTED** [review]

File: test/assets/input/textarea.html (hunk 7 above)
Comment (JoelEinbinder) [reply test/assets/input/textarea.html]: Done.

File: lib/Page.js (hunk 3 above)
Comment (JoelEinbinder) [reply lib/Page.js]: Done.

File: test/assets/input/textarea.html (hunk 1 above)
Comment (JoelEinbinder) [reply test/assets/input/textarea.html]: Done.

File: lib/Page.js (hunk 5 above)
Comment (JoelEinbinder) [reply lib/Page.js]: Done

File: lib/Page.js (hunk 6 above)
Comment (JoelEinbinder) [reply lib/Page.js]: Done.

**Review by aslushnikov — APPROVED** [review]

**Pull request merged**